try.go

```diff
@@ -72,33 +71,76 @@
 //
 package try
 
+import (
+	"fmt"
+	"runtime"
+)
+
 // wrapError wraps an error to ensure that we only recover from errors
 // panicked by this package.
-type wrapError struct{ error }
+type wrapError struct {
+	error
+	frame runtime.Frame
+}
 
 // Unwrap primarily exists for testing purposes.
-func (e wrapError) Unwrap() error { return e.error }
+func (e wrapError) Unwrap() error {
+	return e.error
+}
 
-// Catch catches a previously panicked error and stores it into err.
-// If it successfully catches an error, it calls any provided handlers.
-func Catch(err *error, handlers ...func()) {
-	switch ex := recover().(type) {
+func r(recovered any, fn func(err error, frame runtime.Frame)) {
+	switch ex := recovered.(type) {
 	case nil:
-		return
 	case wrapError:
-		*err = ex.error
-		for _, handler := range handlers {
-			handler()
-		}
+		fn(ex.error, ex.frame)
 	default:
 		panic(ex)
 	}
 }
 
+// Recover recovers an error previously panicked with an E function.
+// If it recovers an error, it calls fn with the error and the runtime frame in which it occurred.
+func Recover(fn func(err error, frame runtime.Frame)) {
+	r(recover(), fn)
+}
+
+// Handle recovers an error previously panicked with an E function and stores it into errptr.
+func Handle(errptr *error) {
+	r(recover(), func(err error, _ runtime.Frame) {
+		*errptr = err
+	})
+}
+
+// HandleF recovers an error previously panicked with an E function and stores it into errptr.
+// If it recovers an error, it calls fn.
+func HandleF(errptr *error, fn func()) {
+	r(recover(), func(err error, _ runtime.Frame) {
+		*errptr = err
+		if err != nil {
+			fn()
+		}
+	})
+}
+
+// F recovers an error previously panicked with an E function, wraps it, and passes it to fn.
+// The wrapping includes the file and line of the runtime frame in which it occurred.
+// F pairs well with testing.TB.Fatal and log.Fatal.
+func F(fn func(...any)) {
+	r(recover(), func(err error, frame runtime.Frame) {
+		fn(fmt.Errorf("%s:%d %w", frame.File, frame.Line, err))
+	})
+}
+
 // E panics if err is non-nil.
 func E(err error) {
 	if err != nil {
-		panic(wrapError{err})
+		pc := make([]uintptr, 1)
+		// 2: runtime.Callers, E
```

Comment:
What about the frame for `E1`?

Reply:
We could rename this as just `e` and then have:

```go
func E(err error) {
	e(err)
}

func E1[A any](a A, err error) A {
	e(err)
	return a
}
```

That way `e` is called from the same stack level in all cases.

Reply:
Good catch. Fixed, with a test.